libraries/common/array.effekt

```diff
@@ -185,6 +209,282 @@ def sum(list: Array[Int]): Int = {
+/// Length is minimum of input lengths.
+///
+/// O(min(N, M))
+def zipWith[A, B, C](this: Array[A], other: Array[B]) { f: (A, B) => C }: Array[C] = {
+  val len = min(this.size, other.size)
+  val result = allocate[C](len)
+  each(0, len) { i =>
+    result.unsafeSet(i, f(this.unsafeGet(i), other.unsafeGet(i)))
+  }
+  result
+}
+
+/// Get the _first_ index of the last occurance of `elem`, if there is any.
+/// 
+/// O(N)
+def indexOf[A](arr: Array[A], toFind: A) { eq: (A, A) => Bool }: Option[Int] = {
+  var found = None[Int]()
+  arr.foreachIndex { (idx, elem) =>
+    if (eq(elem, toFind) && found.isEmpty) {
+      found = Some(idx)
+    }
+  }
+  found
+}
+
+/// Get the _last_ index of the last occurance of `elem`, if there is any.
+/// 
+/// O(N)
+def lastIndexOf[A](arr: Array[A], toFind: A) { eq: (A, A) => Bool }: Option[Int] = {
+  var lastFound = None[Int]()
+  arr.foreachIndex { (idx, elem) =>
+    if (eq(elem, toFind)) {
+      lastFound = Some(idx)
+    }
+  }
+  lastFound
+}
+
+/// Perform binary search to find the index of a given element.
+/// 
+/// O(log N)
+def find[A](arr: Array[A], toFind: A) { ord: (A, A) => Ordering }: Option[Int] = {
+  def binarySearch(lowerIdx: Int, upperIdx: Int): Option[Int] = {
+    if (lowerIdx > upperIdx) None()
+    else {
+      val mid = lowerIdx + (upperIdx - lowerIdx) / 2
+      val current = arr.unsafeGet(mid)
+      ord(current, toFind) match {
+        case Less() => binarySearch(mid + 1, upperIdx)
+        case Greater() => binarySearch(lowerIdx, mid - 1)
+        case Equal() => Some(mid)
+      }
+    }
+  }
+  binarySearch(0, arr.size - 1)
+}
+
+/// Sort array in-place using provided comparison function.
+/// Uses quicksort for O(n log n) average case.
+/// Note: sort is unstable (unlike List.sortBy)
+///
+/// O(N log N) average case
+def sort[A](arr: Array[A]) { ord: (A, A) => Ordering }: Unit = {
+  def unsafeSwap[A](arr: Array[A], i: Int, j: Int): Unit = {
+    val tmp = arr.unsafeGet(i)
+    arr.unsafeSet(i, arr.unsafeGet(j))
+    arr.unsafeSet(j, tmp)
+  }
+
+  def partition(low: Int, high: Int): Int = {
+    val pivot = arr.unsafeGet(high)
+    var i = low - 1
+    
+    each(low, high) { j =>
+      ord(arr.unsafeGet(j), pivot) match {
+        case Less() =>
+          i = i + 1
+          unsafeSwap(arr, i, j)
+        case _ => ()
+      }
+    }
+    unsafeSwap(arr, i + 1, high)
+    i + 1
+  }
+  
+  def quickSort(low: Int, high: Int): Unit = {
+    if (low < high) {
+      val par = partition(low, high)
+      quickSort(low, par - 1)
+      quickSort(par + 1, high)
+    }
+  }
+  
+  quickSort(0, arr.size - 1)
+}
+
+/// swap the value at index i with that at index j
+def swap[A](arr: Array[A], i: Int, j: Int): Unit / Exception[OutOfBounds] = {
+  if (i < 0 || i >= arr.size || j < 0 || j >= arr.size) do raise(OutOfBounds(), "Index out of bounds")
+  val tmp = arr.unsafeGet(i)
+  arr.unsafeSet(i, arr.unsafeGet(j))
+  arr.unsafeSet(j, tmp)
+}
```

Comment:
👀 Somewhat duplicated with the `unsafeSwap` defined locally in `sort`.
Would it be worth it to extract that one out to the outer scope and then define `swap` similarly to how `set` and `get` are defined via their unsafe variants?